akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1148,6 +1159,20 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
       }
       internalInvoke(event)
     }
+
+    // external call
+    def cancel(): Unit = {
+      currentState.get() match {
+        case Completed ⇒
+        case previous ⇒
+          if (!currentState.compareAndSet(previous, Completed)) cancel()
+          else {
+            releaseCallback(this)
+            currentState.set(Completed)
+          }
+      }
+
+    }
```

Comment:
The transitions of the `currentState` are rather complicated and we must be sure that we don't break something there in case `cancel()` is called before `Initialized` so that some pending are silently lost. I don't know if this is a problem, but raising it as warning that might not be immediately visible when reviewing this.

Reply:
An alternative is to just not touch the `currentState` and only remove the `AsyncCallback` from the set. I'll think a bit about what that would lead to.

Reply:
After thinking about it, I'll take care of the pending case instead. I think allowing `invokeWithFeedback` after `cancel` would not behave nicely.